### PR TITLE
perf: Performance Optimization

### LINKED ISSUE #8: Poor performance
The shiny new particle-based weather looks awesome! Thanks again @RatMoleRat :-)

Its performance isn't the best yet, might be mostly due to particle system issues, definitely need some overall improvements there. But since it heavily affects weather I wanted to have this issue here as well to help point it out.

It may also be possible to improve performance here, although conversely it could also be made worse by applying more particles for a better effect. For instance at present the particles start a relatively short distance above the players head, more particles would help but certainly could hit performance further.

Maybe eventually the system could be optimized better by using larger multi-particle texture-based particles or something. But that again sounds more like a fundamental particle system than weather-specific

https://cdn.discordapp.com/attachments/270264625419911192/541435085287850005/unknown.png

### PR DESCRIPTION
**Pull request by kBlaszczyk**
This PR addresses the module's performance issues (fixes #8).

Instead of using a particle pool for every raindrop emitter, we are now using a single particle pool, shared among all raindrop emitters.
Lots of new emitters have been allocated every frame the player moved. Now they are created once and follow the player around passively by being attached to the player entity as its children.

Allocation only happens in case the weather changes, so most of the time the module should be allocation free. The next thing that needs to be improved is the rendering of particle pools in the engine. (partilces in a single pool should be rendered at once)

# Test
Start a new game with the WeatherManager module activated. It should rain and the performance should noticeably better than before.

This PR depends on MovingBlocks/Terasology [#3939](https://github.com/MovingBlocks/Terasology/pull/3939)

### CODE REVIEW

File: src/main/java/org/terasology/weatherManager/systems/EmitWeatherParticleSystem.java [hunk 1]
```diff
@@ -53,19 +60,16 @@
...
-    private Map<Vector2f, EntityRef> particleSpawners;
+    private static final Random random = new FastRandom();
 
-    private Map<Vector2f, EntityBuilder> builders;
-
-    private Map<EntityRef, Vector3f> previousLocations;
+    private String prefabName = SUN;
 
-    private boolean particlesMade;
+    private final Map<EntityRef, List<EntityRef>> emittersByParentEntity = new HashMap<>();
```
Comment (kBlaszczyk) [inline src/main/java/org/terasology/weatherManager/systems/EmitWeatherParticleSystem.java:72]: Note that the module supports creating weather particles for multiple entities.
I kept the functionality, but I don't think it makes any sense. To me, a weather effect is purely local - every player has his own personal rain cloud surrounding him. There is no need to create a rain effect for multiple entities on the same client.
Behavior like this could be useful for attaching rain drop emitters to clouds for example, but that's certainly not a capability of this module and would probably deserve it's own.
I propose to get rid of this map and the multi-entity-support throughout this class. That would simplify things further.
I'd like to hear some opinions in this regard.

File: src/main/java/org/terasology/weatherManager/systems/EmitWeatherParticleSystem.java [hunk 2]
```diff
@@ -366,70 +242,57 @@ private void makeNewParticleEmitters(EntityRef player) {
...
+                    emitterBuilder.setPersistent(false);
 
-            for (EntityRef ref : previousLocations.keySet()) {
-                makeNewParticleEmitters(ref);
+                    if (particlePool != null)
+                        emitterBuilder.getComponent(ParticleEmitterComponent.class).particlePool = particlePool;
+
+                    EntityRef emitter = emitterBuilder.build();
+                    if (particlePool == null)
+                        particlePool = emitter.getComponent(ParticleEmitterComponent.class).particlePool;
```
Comment (kBlaszczyk) [inline src/main/java/org/terasology/weatherManager/systems/EmitWeatherParticleSystem.java:277]: The way I'm reusing the particle pool is pretty ugly. I have one generated by the engine for the first `ParticleEmitterComponent` and extract it again to use it for the others.
It would be much cleaner to just create a new one within the module, but using the class results in a `NoClassDefFoundError`. I think that's weird, considering that the engine is an "implementation" dependency of the module.
Shouldn't the class be usable? Is it intentionally not usable?

**Comment by pollend** [comment]
there seems to be fewer rain particles but the 60fps is a significant improvement.

before: 
image

after: 
image

File: assets/prefabs/particleEffects/rain.prefab [hunk 3]
```diff
@@ -4,8 +4,8 @@
     "texture": "WeatherManager:rain"
   },
   "energyRangeGenerator": {
-    "minEnergy": 45,
-    "maxEnergy": 45
+    "minEnergy": 3,
```
Comment (skaldarnar) [inline assets/prefabs/particleEffects/rain.prefab:7]: @pollend mentions the particles moving slower, is this change related to that? Same for _hail_ in the file above.

File: assets/prefabs/particleEffects/snow.prefab [hunk 4]
```diff
@@ -4,8 +4,8 @@
     "texture": "WeatherManager:snow"
   },
   "energyRangeGenerator": {
-    "minEnergy": 45,
-    "maxEnergy": 45
+    "minEnergy": 3,
```
Comment (skaldarnar) [inline assets/prefabs/particleEffects/snow.prefab:7]: well, snow should probably fall slower, so that change might be okay? I don't know ...

File: src/main/java/org/terasology/weatherManager/systems/EmitWeatherParticleSystem.java (hunk 1 above)
Comment (skaldarnar) [reply src/main/java/org/terasology/weatherManager/systems/EmitWeatherParticleSystem.java:72]: How would we model a distant thunderstorm for a player? Or the player standing in the rain while a snow storm approaches them?

These questions are not to oppose your suggestion but to get an idea whether these scenarios would work with the map but won't be possible without...

Otherwise, I agree that this might be a bit more complicated that actually needed for a client system.

File: src/main/java/org/terasology/weatherManager/systems/EmitWeatherParticleSystem.java [hunk 5]
```diff
@@ -53,19 +60,16 @@
...
 
-    private String prefabName = SUN;
-
-    private Map<Vector2f, EntityRef> particleSpawners;
+    private static final Random random = new FastRandom();
 
-    private Map<Vector2f, EntityBuilder> builders;
-
-    private Map<EntityRef, Vector3f> previousLocations;
+    private String prefabName = SUN;
```
Comment (skaldarnar) [inline src/main/java/org/terasology/weatherManager/systems/EmitWeatherParticleSystem.java:70]: Suggestion: Use a case insensitive `Name` here instead of a plain String.

Comment (skaldarnar) [reply src/main/java/org/terasology/weatherManager/systems/EmitWeatherParticleSystem.java:70]: Also, would `currentWeather` be a better name than `prefabName`?

File: src/main/java/org/terasology/weatherManager/systems/EmitWeatherParticleSystem.java [hunk 6]
```diff
@@ -182,180 +179,59 @@ public void onStartHailEvent(StartHailEvent event, EntityRef worldEntity) {
...
-                    boolean isSnow = ref.getParentPrefab().getName().equalsIgnoreCase("WeatherManager:snow");
-                    boolean isHail = ref.getParentPrefab().getName().equalsIgnoreCase("WeatherManager:hail");
-                    if (isRain || isSnow || isHail) {
-                        ref.destroy();
-                    }
-                }
+        if (weather != null && weatherManagerSystem.getCurrentSeverity() != null) {
+            maxDownfall = -7;
+            minDownfall = -4;
+            switch (weatherManagerSystem.getCurrentSeverity()) {
```
Comment (skaldarnar) [inline src/main/java/org/terasology/weatherManager/systems/EmitWeatherParticleSystem.java:209]: what are these numbers, what do they express? And why is the default above even lighter than `LIGHT`?

File: src/main/java/org/terasology/weatherManager/systems/EmitWeatherParticleSystem.java [hunk 7]
```diff
@@ -366,70 +242,57 @@ private void makeNewParticleEmitters(EntityRef player) {
...
-                    break;
-            }
+                    emitterBuilder.getComponent(VelocityRangeGeneratorComponent.class).minVelocity.set(minVelocity);
+                    emitterBuilder.getComponent(VelocityRangeGeneratorComponent.class).maxVelocity.set(maxVelocity);
+                    emitterBuilder.getComponent(LocationComponent.class).setWorldPosition(emitterPosition);
+                    emitterBuilder.setPersistent(false);
 
-            for (EntityRef ref : previousLocations.keySet()) {
-                makeNewParticleEmitters(ref);
+                    if (particlePool != null)
```
Comment (skaldarnar) [inline src/main/java/org/terasology/weatherManager/systems/EmitWeatherParticleSystem.java:272]: Won't this always be `null` at this point, or am I missing something?

File: src/main/java/org/terasology/weatherManager/systems/EmitWeatherParticleSystem.java (hunk 2 above)
Comment (skaldarnar) [reply src/main/java/org/terasology/weatherManager/systems/EmitWeatherParticleSystem.java:277]: Not sure, but maybe we need to mark the implementation as `@API` in the engine to be accessible by modules? 🤔

File: src/main/java/org/terasology/weatherManager/systems/WeatherManagerSystem.java [hunk 8]
```diff
@@ -160,11 +158,12 @@ private void changeWeather(ConditionAndDuration conditionAndDuration) {
     @ReceiveEvent
     public void onTimeEvent(DelayedActionTriggeredEvent event, EntityRef worldEntity) {
 
-        current = weatherConditionProvider.getNext();
-
-        currentWeather = current.condition.downfallCondition.getDownfallValues().type;
-        severity = current.condition.downfallCondition.getDownfallValues().amount;
-        currentWind = current.condition.wind;
+        if (event.getActionId().equals("RandomWeather")) {
```
Comment (skaldarnar) [inline src/main/java/org/terasology/weatherManager/systems/WeatherManagerSystem.java:161]: Why did this become necessary? And are there other event ids that can be used here? Also, I don't like the use of plain strings here, but I guess that's how the event ids work... 🤷

File: assets/prefabs/particleEffects/rain.prefab (hunk 3 above)
Comment (kBlaszczyk) [reply assets/prefabs/particleEffects/rain.prefab:7]: No, the energy represents their life time. Reducing it down makes sure they won't live for too long, since the particle collision detection is quite unreliable.

File: assets/prefabs/particleEffects/snow.prefab (hunk 4 above)
Comment (kBlaszczyk) [reply assets/prefabs/particleEffects/snow.prefab:7]: The fall speed is hard coded in `EmitWeatherParticleSystem`. I haven't changed the values, but they differ depending on the weather severity (heavy, moderate and light - specified randomly)

File: src/main/java/org/terasology/weatherManager/systems/EmitWeatherParticleSystem.java (hunk 6 above)
Comment (kBlaszczyk) [reply src/main/java/org/terasology/weatherManager/systems/EmitWeatherParticleSystem.java:209]: That's what I'm referring to in the other comment. It's the definition of the particles fall speed. Currently the only "other" `Severity` is NONE. I assume it is supposed to be weaker than LIGHT.

File: src/main/java/org/terasology/weatherManager/systems/EmitWeatherParticleSystem.java (hunk 7 above)
Comment (kBlaszczyk) [reply src/main/java/org/terasology/weatherManager/systems/EmitWeatherParticleSystem.java:272]: It won't be `null` after the first iteration. `particlePool` is being initialized a few lines below.

File: src/main/java/org/terasology/weatherManager/systems/WeatherManagerSystem.java (hunk 8 above)
Comment (kBlaszczyk) [reply src/main/java/org/terasology/weatherManager/systems/WeatherManagerSystem.java:161]: Setting the weather using the console commands didn't work.
When executing one of the commands, the `WeatherManagerSystem` would change the weather to the specified target directly - and also schedule a delayed action to just change the weather. This weather change was always random, which resulted in the weather being set to the desired one and then immediately changed to a random one.
I fixed the `onTimeEvent()` logic by differentiating between specific and random weather changes.
And yes, I assume that's how event IDs are supposed to be used.

**Comment by kBlaszczyk** [comment]
> there seems to be fewer rain particles but the 60fps is a significant improvement. the seem to be also moving a lot slower so the behavior seems to differ quite a bit.

This implementation is using 100 particle emitters, the number can be tweaked using `PARTICLE_EMITTERS_COUNT`. The size of the particle area should probably be increased as well. In case you are running against the wind, you sometimes outrun the falling particles.
The fall speed _should_ be identical. My guess is that you compared light rain in the new implementation with heavy rain in the old one, but I'll check as well, if I notice a difference.

File: src/main/java/org/terasology/weatherManager/systems/EmitWeatherParticleSystem.java (hunk 1 above)
Comment (kBlaszczyk) [reply src/main/java/org/terasology/weatherManager/systems/EmitWeatherParticleSystem.java:72]: Interesting idea.
But the current event API doesn't even support creating a specific weather effect for a specific entity.
It just passively creates these weather effects for every spawning or connecting player (which is completely unnecesarry on a single client). And then changes the weather for all entities equally.
In conclusion, if these kind of changes are supposed to be implemented at any time, the API and implementation need to be changed quite a bit. One might as well introduce the map in question when it becomes required. Right now I think this mechanism does more harm than good.

File: src/main/java/org/terasology/weatherManager/systems/EmitWeatherParticleSystem.java (hunk 5 above)
Comment (kBlaszczyk) [reply src/main/java/org/terasology/weatherManager/systems/EmitWeatherParticleSystem.java]: done

File: src/main/java/org/terasology/weatherManager/systems/EmitWeatherParticleSystem.java (hunk 2 above)
Comment (kBlaszczyk) [reply src/main/java/org/terasology/weatherManager/systems/EmitWeatherParticleSystem.java:278]: That works. Do you think I should add that to the other PR?

File: src/main/java/org/terasology/weatherManager/systems/EmitWeatherParticleSystem.java (hunk 1 above)
Comment (skaldarnar) [reply src/main/java/org/terasology/weatherManager/systems/EmitWeatherParticleSystem.java:73]: In that case, please go ahead and simplify it, either in this PR or in a follow up 😉

Comment (kBlaszczyk) [reply src/main/java/org/terasology/weatherManager/systems/EmitWeatherParticleSystem.java:72]: Alright, I'll do it in this one.

**Comment by kBlaszczyk** [comment]
@skaldarnar 
The effect will now exclusively affect the local player. I'm quite happy with the usage of `LocalPlayer` to achieve this. Things got cleaned quite a bit.

File: src/main/java/org/terasology/weatherManager/systems/EmitWeatherParticleSystem.java [hunk 9]
```diff
@@ -80,51 +76,40 @@
...
-    @ReceiveEvent
-    public void playerRespawned(OnPlayerRespawnedEvent event, EntityRef player) {
-        begin(player);
-    }
-
     @ReceiveEvent(components = ClientComponent.class)
     public void onConnect(ConnectedEvent connected, EntityRef entity) {
-        begin(entity);
+        if (localPlayer.getId() == 0)
+            localPlayer = entity;
```
Comment (skaldarnar) [inline src/main/java/org/terasology/weatherManager/systems/EmitWeatherParticleSystem.java]: We should get some style checker running for modules again 🧐 please do _always_ use braces with if-statements. 
```suggestion
        if (localPlayer.getId() == 0) {
            localPlayer = entity;
        }
```

Comment (skaldarnar) [reply src/main/java/org/terasology/weatherManager/systems/EmitWeatherParticleSystem.java]: same for the other occurrences

Comment (kBlaszczyk) [reply src/main/java/org/terasology/weatherManager/systems/EmitWeatherParticleSystem.java]: alright
